app/models/enterprise.rb

```diff
@@ -91,15 +91,15 @@ class Enterprise < ActiveRecord::Base
     joins(:shipping_methods).
       joins(:payment_methods).
       merge(Spree::PaymentMethod.available).
-      select('DISTINCT enterprises.*')
+      select('enterprises.id')
```

Comment:
But this is not equivalent. Where we were getting all columns, now we just get the id.

Reply:
the * is not allowed in rails 4 and ready_for_checkout is only used for entities in EnterpriseInjectionData.active_distributors.
I have tested and the method works perfectly fine and the enterprises are loaded correctly.
 https://github.com/openfoodfoundation/openfoodnetwork/blob/a45782f59b9f355c6b7649554496739f3c29abe9/lib/open_food_network/enterprise_injection_data.rb#L3

so I think we are good here.

Reply:
Changing the `SELECT` to just use `enterprises.id` is misleading IMO because the scope name does not clarify that only the ID would be included.

This only works because `Enterprise.distributors_with_active_order_cycles` still uses `select("DISTINCT enterprises.*")` (see [here](https://github.com/openfoodfoundation/openfoodnetwork/blob/0f461067ab763ae766521b9f132a30008a37e446/app/models/enterprise.rb#L132)), so the resulting `SELECT` is `DISTINCT enterprises.*, enterprises.id`.

In Rails 4. there is [`ActiveRecord::Associations::CollectionProxy#distinct`](https://api.rubyonrails.org/v4.2.11.1/classes/ActiveRecord/Associations/CollectionProxy.html#method-i-distinct): (`v4.2.11`)

```
> puts User.joins(:recipes).distinct.to_sql
SELECT DISTINCT "users".* FROM "users" INNER JOIN "recipes" ON "recipes"."user_id" = "users"."id"
```

If we can't change `distributors_with_active_order_cycles` to not to use `DISTINCT enterprises.*`, I think we should leave this particular line as is for now and just wait for the Rails 4 upgrade before switching to `distinct`.

Reply:
thanks @kristinalim

"and just wait for the Rails 4 upgrade"
this "select distinct enterprises.*" doesnt work in rails 4. all the changes in this PR are fixing broken specs in the spree 2-1 build (running on rails 4). I mean, this is the rails 4 upgrade 🎉 

I am not sure why this "distinct enterprises.*" is breaking stuff and not the one in distributors_with_active_order_cycles (maybe I just missed other broken specs in the 2-1 build related to distributors_with_active_order_cycles).

I think you are suggesting I revert this change until we fix this and the one in distributors_with_active_order_cycles.

what if I change the name of this scope to be more appropriate ready_for_checkout_ids?

let me know, I can either revert this or rename the scope if you think it's acceptable (I prefer to keep this because this fixes lots of tests in the 2-1 build).

thanks!

Reply:
Ah yes, nice. :slightly_smiling_face:

> what if I change the name of this scope to be more appropriate ready_for_checkout_ids

I did a quick check and I do not understand either. :grinning: It's not ideal, but okay - that sounds good to me.

Reply:
ok, I think it was my misconception that select * was not supported in rails 4. There's nothing wrong with the select * in rails 4

(there is something very wrong with "select *" in my mind, "select *" is an anti pattern, it should always be avoided, the specific fields should always be listed, but I'll ignore this case for now).

So, I went back to v2.1. to verify and what is wrong in the not_ready_for_checkout scope:
https://github.com/openfoodfoundation/openfoodnetwork/blob/98d1cbec8c7a0f1b71ed16e2e527f250bd8fe8b6/app/models/enterprise.rb#L101
the problem is that in rails 4 we can't do "id IN (scope)" where the scope does select *, we need to select id.

So, I added a new commit to this PR to make ready_enterprises select the id field from the ready_for_checkout scope: [6091b1c319811c95e9adc4221562b21389111d30](https://github.com/openfoodfoundation/openfoodnetwork/pull/3709/commits/6091b1c319811c95e9adc4221562b21389111d30)
two points about the commit:
- I checked and the select statement does not trigger additional queries
- the enterprises.* is needed otherwise we get a "ambiguous column" (like in other changes in this PR)

Let me know if this makes sense for you :-)
Thanks!

Reply:
The commit looks great. :smile: